ethcore/light/src/transaction_queue.rs

```diff
@@ -0,0 +1,421 @@
+					// don't accept txs from before known current nonce.
+					if acct_txs.cur_nonce.is_known() {
+						return Err(TransactionError::Old)
+					}
+
+					// lower our assumption until corrected later.
+					acct_txs.cur_nonce = CurrentNonce::Assumed(nonce);
+				}
+
+				match acct_txs.current.binary_search_by(|x| x.nonce.cmp(&nonce)) {
+					Ok(idx) => {
+						trace!(target: "txqueue", "Replacing existing transaction from {} with nonce {}",
+							sender, nonce);
+
+						acct_txs.current[idx] = tx.clone();
+
+						TransactionImportResult::Current
+					}
+					Err(idx) => {
+						let cur_len = acct_txs.current.len();
+						let incr_nonce = nonce + 1.into();
+
+						// current is sorted with one tx per nonce,
+						// so if a tx with given nonce wasn't found that means it is either
+						// earlier in nonce than all other "current" transactions or later.
+						debug_assert!(idx == 0 || idx == cur_len);
+
+						if idx == 0 && acct_txs.current.first().map_or(false, |f| f.nonce != incr_nonce) {
+							let old_cur = ::std::mem::replace(&mut acct_txs.current, vec![tx.clone()]);
+
+							trace!(target: "txqueue", "Moving {} transactions with nonce > {} to future",
+								old_cur.len(), incr_nonce);
+
+							for future in old_cur {
+								let future_nonce = future.nonce;
+								acct_txs.future.insert(future_nonce, future);
+							}
+
+							TransactionImportResult::Current
+						} else if idx == cur_len && acct_txs.current.last().map_or(false, |f| f.nonce + 1.into() != nonce) {
+							trace!(target: "txqueue", "Queued future transaction for {}, nonce={}", sender, nonce);
+							let future_nonce = nonce;
+							acct_txs.future.insert(future_nonce, tx.clone());
+
+							TransactionImportResult::Future
+						} else {
+							trace!(target: "txqueue", "Queued current transaction for {}, nonce={}", sender, nonce);
+
+							// insert, then check if we've filled any gaps.
+							acct_txs.current.insert(idx, tx.clone());
+							acct_txs.adjust_future();
+
+							TransactionImportResult::Current
+						}
+					}
+				}
+			}
+		};
+
+		self.by_hash.insert(hash, tx);
+		Ok(res)
+	}
+
+	/// Get pending transaction by hash.
+	pub fn transaction(&self, hash: &H256) -> Option<SignedTransaction> {
+		self.by_hash.get(hash).map(|tx| (&**tx).clone())
+	}
+
+	/// Get the next nonce for a given address based on what's within the queue.
+	/// If the address has no queued transactions, then `None` will be returned
+	/// and the next nonce will have to be deduced via other means.
+	pub fn next_nonce(&self, address: &Address) -> Option<U256> {
+		self.by_account.get(address).map(AccountTransactions::next_nonce)
+	}
+
+	/// Get all transactions ready to be propagated.
+	/// `best_block_number` and `best_block_timestamp` are used to filter out conditionally
+	/// propagated transactions.
+	pub fn ready_transactions(&self, best_block_number: u64, best_block_timestamp: u64) -> Vec<PendingTransaction> {
```

Comment:
Haven't implemented the transaction propagation code for the light node just yet. I'll add docs that it's in "arbitrary" order.